Add a polynomial dialect shell, attributes, and types

j2kun opened the pull request:
RFC: https://discourse.llvm.org/t/rfc-a-poly-dialect-for-polynomial-arithmetic/73891

This PR implements the minimal work needed to represent the polynomial type such that it can be tested with `lit`.

Please let me know if this PR is too big to review, and I can split it into smaller pieces:

- Dialect shell
- `Polynomial` storage type
- Polynomial attributes (`PolynomialAttr` and `RingAttr`)
- `polynomial.polynomial` type

Note that I also have C++ unit tests for the polynomial storage type that can be ported from https://github.com/google/heir/blob/main/lib/Dialect/Polynomial/IR/PolynomialTest.cpp, if desired. I believe C++ unit tests are considered less preferred than `lit` tests, but let me know if you prefer unit tests on top of lit tests.

j2kun committed:
polynomial: add dialect shell, attrs, and types

ftynse commented on mlir/lib/Dialect/Polynomial/IR/PolynomialDetail.h:
```diff
@@ -0,0 +1,65 @@
+
+  unsigned numTerms;
+
+  MLIRContext *context;
```
Why is this even needed? Doesn't seem to be used.

ftynse replied:
> As I understand, nothing outside the impl should be stored on the Polynomial class...

Ack.

I suppose my question is subsumed by Mehdi's question below about the difference between Polynomial and PolynomialAttr. I'll comment there.

ftynse commented on mlir/test/Dialect/Polynomial/types.td:
Please add some more positive tests, at least for the printing and parsing of the polynomial attribute.

Also, please add tests for all user-visible diagnostics such as negative tests for parsing.

j2kun replied:
Done.

- 115b17ebf3c6 has diagnostic tests
- 0b8d0ca811e7 has more positive tests

joker-eph reviewed (commented):
Thanks for the contribution!

(I started reviewing, but I see Alex sent comments so just flushing a few...)

j2kun committed:
fix include guards

j2kun committed:
add top level docs to Monomial

j2kun committed:
use /// for top level comment

j2kun committed:
make hash_value arg const

j2kun committed:
std::string -> StringRef

j2kun committed:
update/align header comments

j2kun committed:
remove tex quotes

j2kun committed:
use auto over OwningArrayRef

j2kun committed:
use SmallVector instead of std::vector, clarify docstring

j2kun committed:
Fix parser error string style

j2kun committed:
small string size 512 -> 16, and snake_case to camelCase

j2kun committed:
comment capitalization style

j2kun committed:
use consistent verbose attribute property names

j2kun committed:
std::vector -> SmallVector

j2kun committed:
pass variable stringref by reference

j2kun committed:
use llvm::is_contained

j2kun committed:
fix emitError string concatenation

j2kun committed:
remove stray comment

j2kun committed:
remove unnecessary llvm:: prefix

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,157 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef INCLUDE_MLIR_DIALECT_POLYNOMIAL_IR_POLYNOMIAL_H_
```
Fixed by 51cc8c4a0bf9ed87e467cefd0ba2ea59b9ef47cb

j2kun replied:
I will tentatively mark "resolve" on this and other threads I believe I have addressed. Please reopen if you feel it's not adequately resolved.

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,157 @@
+struct PolynomialStorage;
+} // namespace detail
+
+class Monomial {
```
Fixed by 8116afa66f61d42d049afc62f6eea2f2c8fe2b2e

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,157 @@
+
+namespace polynomial {
+
+// This restricts statically defined polynomials to have at most 64-bit
```
Fixed by 58b834d59972aa3eddaa0eddf492f9f1a54eb121

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,157 @@
+
+  // Prints polynomial to 'os'.
+  void print(raw_ostream &os) const;
+  void print(raw_ostream &os, const std::string &separator,
```
Fixed by b9a89604123ba87c63d6cb2918975a6db1da4ed3

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/PolynomialAttributes.h:
```diff
@@ -0,0 +1,18 @@
+//===- PolynomialAttributes.h - Attributes for the Polynomial dialect -*- C++
+//-*-===//
```
Fixed by dc9dcaef3152c5b3d6afe73cdf3ee2fb0d848a4

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/PolynomialAttributes.td:
```diff
@@ -0,0 +1,81 @@
+    A ring describes the domain in which polynomial arithmetic occurs. The ring
+    attribute in `polynomial` represents the more specific case of polynomials
+    with a single indeterminate; whose coefficients can be represented by
+    another MLIR type (`ctype`); and, if the coefficient type is integral,
```
I will go ahead and make them all the verbose version. Fixed in 2cec095f91ce849a5673fe7accf75996e7fe9371

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/PolynomialAttributes.td:
````diff
@@ -0,0 +1,81 @@
+    %0 = ... : polynomial.polynomial<#ring>
+    ```
+
+    In this case, the value of a polynomial is always ``converted'' to a
````
My mistake, this is just because I write lots of LaTeX in other contexts and have picked up this bad habit of using tex quotes where they don't belong.

Fixed in 131557f262037f5290a22be13e6b9be0ecd827b0

j2kun commented on mlir/lib/Dialect/Polynomial/IR/Polynomial.cpp:
```diff
@@ -0,0 +1,97 @@
+  };
+
+  // A polynomial's terms are canonically stored in order of increasing degree.
+  llvm::OwningArrayRef<Monomial> monomialsCopy =
```
Fixed by b342feda22137224d2da967e632179b77ddf25c1

j2kun commented on mlir/lib/Dialect/Polynomial/IR/Polynomial.cpp:
```diff
@@ -0,0 +1,97 @@
+
+Polynomial Polynomial::fromCoefficients(ArrayRef<int64_t> coeffs,
+                                        MLIRContext *context) {
+  std::vector<Monomial> monomials;
```
I need the index to determine the exponent in this case, so `map_to_vector` doesn't work (though TIL it exists, thanks!). Also clarified the documentation on this function to explain the implicit exponent.

Fixed in db0ae3c9973987e91e40e5db8e7b276204c9159d

j2kun commented on mlir/lib/Dialect/Polynomial/IR/Polynomial.cpp:
```diff
@@ -0,0 +1,97 @@
+    if (term.coefficient == 1 && term.exponent.uge(1)) {
+      coeffToPrint = "";
+    } else {
+      llvm::SmallString<512> coeffString;
```
Changed to 16 characters in 758d4f098c91af75c5a8b019a070cf8bc14b5262

ftynse replied:
Did you forget to push this?

j2kun replied:
I must have... re-fixed in 6c2526dbc74

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+  if (result.has_value()) {
+    if (failed(*result)) {
+      parser.emitError(parser.getCurrentLocation(),
+                       "Invalid integer coefficient.");
```
Fixed in da6bba7a53853fa25743e53b28041d1a247fcd93

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+  // Variable name
+  result = parser.parseOptionalKeyword(variable);
+  if (!result.has_value() || failed(*result)) {
+    // we allow "failed" because it triggers when the next token is a +,
```
Fixed in b10746db30b1df35ae462e646ffc24fc7b4949b2

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+  if (failed(parser.parseLess()))
+    return {};
+
+  std::vector<Monomial> monomials;
```
Fixed in c5120fd76a1960e221b2a486ea924c1e064e5ef6

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+/// `monomial` with the results, and the `variable` outparam with the parsed
+/// variable name.
+ParseResult parseMonomial(AsmParser &parser, Monomial &monomial,
+                          llvm::StringRef *variable, bool *isConstantTerm) {
```
Note that this is passed through for `parseOptionalKeyword`, which takes a `StringRef *`, so at best I could move the `&` unary op down one level to the call to `parseOptionalKeyword`.

Fixed in 449ce187b409c19204d2cea7198b33031b4a62b3

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+    return {};
+
+  std::vector<Monomial> monomials;
+  llvm::SmallSet<std::string, 2> variables;
```
I couldn't figure out how to get llvm::join to work with StringSet, when used like

```cpp
llvm::StringSet<> variables;
<...>
std::string vars = llvm::join(variables.begin(), variable.end(), ", ");
```

I get an error like

```
error: no matching function for call to ‘llvm::StringRef::StringRef(llvm::StringMapEntry<std::nullopt_t>&)’
```

OK to leave it as SmallSet?

ftynse replied:
Something like `llvm::join(llvm::make_first_range(stringSet), ", ");` should work.
But I don't have a strong opinion about a data structure here, so okay to keep unless somebody else has good arguments.

j2kun replied:
Fixed in 1b088b3b29636b4a093fe9e7040e36ae0227cefc

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+    }
+    monomials.push_back(parsedMonomial);
+
+    if (exponents.count(parsedMonomial.exponent) > 0) {
```
Fixed in f7cabe44200340c10f5686e1bcb00c67616ee01d

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+      parsedMonomial.exponent.toStringSigned(coeff_string);
+      parser.emitError(parser.getCurrentLocation(),
+                       "At most one monomial may have exponent " +
+                           coeff_string + ", but found multiple.");
```
Fixed in 4c03391bf6d4c927f767ee0363ea5e61b8bccbf9

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+    }
+  }
+
+  // insert necessary constant ops to give as input to extract_slice.
```
Fixed in d8f90f2530f70183be0c6dac7fab8fc8f62271ee

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+    if (failed(parser.parseEqual()))
+      return {};
+
+    IntegerType iType = llvm::dyn_cast<IntegerType>(typeAttr.getValue());
```
Fixed in 66843d67d491e24bedde5e5d37e6fe46ef6d1214

j2kun replied:
Hmmm this broke the build and I tried `using namespace mlir` but it doesn't seem to help. Reverting for now.

ftynse replied:
That's strange. Try including `mlir/Support/LLVM.h` (it is included transitively by most other headers, but not all).

j2kun replied:
That import is already there, but it looks like the static version of dyn_cast was replaced by a member function. Fixed in 6c153e66cd0b0f13b7f5b2566f59ec98eff917cb

j2kun committed:
fiddle with namespaces a bit more

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,157 @@
+///
+/// The symbols used as the polynomial's indeterminate don't matter, so long as
+/// it is used consistently throughout the polynomial.
+class Polynomial {
```
I intend to implement folding for polynomial ops, and this separates the class that would implement the underlying arithmetic from the attribute class. I'm not sure it's strictly necessary, and I can combine the two if you think that would make more sense.

joker-eph replied:
Ah I can see an argument of efficiency to be able to perform arithmetic without recreating / uniquing an attribute on every operation.
Can you document the difference between this and the attribute here?

ftynse replied:
But isn't the `Polynomial` object also uniqued here? This looks similar to `AffineMap`/`AffineMapAttr`, and I am probably missing what the second level of boxing brings us here? (I understand that `Monomial` may be a context-unique object while not being available as an attribute though.)

j2kun replied:
Ah, yes, Polynomial is uniqued. I think this discrepancy is mostly my lack of familiarity with this part of MLIR.

What I originally intended to do was just separate conceptually a polynomial from an attribute (just a habit of treating different things differently), and I was borrowing from AffineMap in building this.

I would be open to combining them or keeping them separate and moving the MLIR Context and uniquing stuff to the attribute. Which is preferred?

ftynse replied:
IIUC, the distinction between `AffineMap` and `AffineMapAttr` is a historic artifact at this point (originally, `AffineMap` was a fundamental part of `MemRefType` and we didn't have fully parameterizable types, let alone types parameterized by an attribute). Maybe there is some aspect that I'm forgetting...

We could have a separate `Polynomial` class, but it doesn't have to be a context-owned or uniqued object. This would be similar to the classes in the Presburger library, which are also related to mathematical concepts.

Let's also hear what @joker-eph thinks.

j2kun replied:
Pinging @joker-eph  for any further thoughts on this.

j2kun replied:
I am returning to this ancient PR after months of distraction. I will go ahead and remove the unique-ing from Polynomial

j2kun replied:
Fixed in 8d7f5d1c978

j2kun committed:
= nullptr

j2kun committed:
use hash_combine

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/PolynomialAttributes.td:
```diff
@@ -0,0 +1,81 @@
+  let parameters = (ins
+    Builtin_TypeAttr: $coefficientType,
+    OptionalParameter<"std::optional<IntegerAttr>">: $coefficientModulus,
+    OptionalParameter<"std::optional<PolynomialAttr>">: $polynomialModulus
```
I was a bit confused on this point. When I use `OptionalParameter` I, perhaps naively, expected it to generate builders with the optional parts missing, or builders that take `std::optional` as input, but instead it seems the default is null. 

The [docs state](https://mlir.llvm.org/docs/DefiningDialects/AttributesAndTypes/#optional-and-default-valued-parameters)

> When using OptionalParameter, the default value is set to the C++ default-constructed value for the C++ storage type. For example, Optional<int> will be set to std::nullopt and Attribute will be set to nullptr. The presence of these parameters is tested by comparing them to their “null” values.

I took this to mean that if I wanted to use the std::optional type in the generated code, I needed to make that explicit as above. Are you implying I should remove std::optional and test the underlying Attributes against `nullptr`? I can do that, I just feel it's less brittle to have the type encode optionality.

joker-eph replied:
I'm just looking at what's the current usage, there are 23 uses with `std::optional` and none are in combination with Attributes (only `std::optional<int>` actually). There are 131 instances without `std::optional` (89 with Attribute).

How do the accessors look like?

j2kun replied:
The generated class definition is

```cpp
class RingAttr : public ::mlir::Attribute::AttrBase<RingAttr, ::mlir::Attribute, detail::RingAttrStorage> {
public:
  using Base::Base;
  static RingAttr get(::mlir::MLIRContext *context, 
                      ::mlir::TypeAttr coefficientType, 
                      std::optional<IntegerAttr> coefficientModulus, 
                      std::optional<PolynomialAttr> polynomialModulus);
  std::optional<IntegerAttr> getCoefficientModulus() const;
  std::optional<PolynomialAttr> getPolynomialModulus() const;
  ...
};
```

This seems preferable to null checking to me, but if you feel strongly I can remove `std::optional`.

j2kun replied:
Pinging @joker-eph  for any further thoughts on this.

joker-eph replied:
Sorry I missed your answer! 

I'm in favor of making it explicit, (I even remember advocating having a wrapper around returned pointers which can be nullptr, something like `maybe_null<T *>` which would need to be checked before dereferenced and adopt the `gsl::not_null<T *>` API otherwise, but I was in minority when I proposed this).

That said I also like keeping the codebase a bit more uniform: I feel this is a choice that should be made in the implementation of `OptionalParameter` and not on a case-by-case basis.

j2kun replied:
Fixed in 8d7f5d1c978

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,157 @@
+}
+
+inline ::llvm::hash_code hash_value(Monomial arg) {
+  return ::llvm::hash_value(arg.coefficient) ^ ::llvm::hash_value(arg.exponent);
```
TIL. Fixed in cb278a510503de89a496227d5fe93b1ee186a094

j2kun commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,157 @@
+  friend ::llvm::hash_code hash_value(Polynomial arg);
+
+private:
+  ImplType *terms{nullptr};
```
Fixed in b75bf1eac151faa19541382a9f28e65b3da05024

j2kun commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,217 @@
+
+  // Variable name
+  result = parser.parseOptionalKeyword(variable);
+  if (!result.has_value() || failed(*result)) {
```
I think that putting a `parseOptionalPlus` in this function, if it succeeds, would consume the `+` token and then the parser would not have it later after returning to `PolynomialAttr::parse` when deciding if there are more monomials to parse (in the current revision, line `PolynomialAttributes.cpp:118`).

Is there a way to parseOptionalPlus and then, if it succeeds, pop the `+` back on the top of the parsing context so that when it returns it can re-parse it?

ftynse replied:
I'm not sure I understand the syntax here. Maybe adding tests, or pasting examples, could help. There's no lookahead exposed at this level of the API.

One specific thing that looks tricky here: if parser returns failure as opposed to nullopt, it means it may have reported an error diagnostic to the user, and it will be very surprising if the parsing procedure actually succeeds after that. It may not be the case right now, though.

j2kun replied:
I tried rewriting this to make it less hacky in 6f175107f6937bb0c2497a80b3de73fa336d0c89, but without being able to look ahead in some manner, I feel it will remain hacky.

The complexity is that a monomial can have the following forms, separated by `+`:

- `4` : a degree-0 term (variable, exponent optional)
- `x`: a coefficient-1, degree-1 term (coefficient, exponent optional)
- `2x` : a degree-1 term (exponent optional)
- `x**5`: a coefficient-1 term (coefficient optional)
- `3x**5`: no parts omitted

Handling all of these possibilities is a bit of spaghetti because if a parse method succeeds, it consumes the token. The options I see to make it better are:

- Simplify the allowed syntax so that every monomial has the form `<coeff> <symbol>**<exponent>`, even if `exponent` is zero or `coeff` is 1. Then there's only one legal path.
- Somehow extend the parser so that I can attempt to parse the different allowed monomial forms in completely separate functions, and then if it fails "reset" to the previous parser location before trying the next one. Then I can work from the most specific form (no parts omitted) to the least specific (degree-0 monomials). E.g., a `parseFullMonomial` function that does the last item in the list above, then fall back to `parseMonomialMissingCoefficient`, then `parseMonomialMissingExponent`, etc.
- Provide a generalized comma-separated list parsing function to a "parse token-separated things" to handle the variadic nature.

j2kun committed:
add diagnostic tests

j2kun committed:
cleanup minor attribute suggestions

j2kun committed:
add more positive tests

j2kun committed:
use StringSet

j2kun committed:
use dyn_cast member function

j2kun committed:
try refactoring monomial parser somewhat

j2kun committed:
don't unique Polynomial, remove std::optional from OptionalParameter

j2kun commented:
@ftynse @joker-eph Sorry for the months-long delay on this. To recap, the main changes since the last review was to remove the uniquing of Polynomial, during which I also simplified the Polynomial class to store its monomial terms as a simple `SmallVector`, rather than use `llvm::TrailingObjects`.

j2kun committed:
clang-format

ftynse commented on mlir/include/mlir/Dialect/Polynomial/IR/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+set(LLVM_TARGET_DEFINITIONS PolynomialOps.td)
+mlir_tablegen(PolynomialDialect.cpp.inc -gen-dialect-defs -dialect=polynomial)
+mlir_tablegen(PolynomialDialect.h.inc -gen-dialect-decls -dialect=polynomial)
+add_public_tablegen_target(MLIRPolynomialDialectIncGen)
```
`add_dependencies(mlir-headers MLIRPolynomialDialectIncGen)` may be needed here.

j2kun replied:
Fixed in 4f8be047edc

ftynse commented on mlir/include/mlir/Dialect/Polynomial/IR/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+mlir_tablegen(PolynomialDialect.h.inc -gen-dialect-decls -dialect=polynomial)
+add_public_tablegen_target(MLIRPolynomialDialectIncGen)
+
+mlir_tablegen(PolynomialAttributes.cpp.inc -gen-attrdef-defs -attrdefs-dialect=polynomial)
+mlir_tablegen(PolynomialAttributes.h.inc -gen-attrdef-decls -attrdefs-dialect=polynomial)
+mlir_tablegen(PolynomialOps.cpp.inc -gen-op-defs)
+mlir_tablegen(PolynomialOps.h.inc -gen-op-decls)
+mlir_tablegen(PolynomialTypes.cpp.inc -gen-typedef-defs -typedefs-dialect=polynomial)
+mlir_tablegen(PolynomialTypes.h.inc -gen-typedef-decls -typedefs-dialect=polynomial)
+add_public_tablegen_target(MLIRPolynomialAttributesIncGen)
+add_public_tablegen_target(MLIRPolynomialOpsIncGen)
+add_public_tablegen_target(MLIRPolynomialTypesIncGen)
```
This looks incorrect. IIUC, tablegen commands are stateful. Specifically, one first has to set `LLVM_TARGET_DEFINITIONS` to point to the source .td, then list generation rules in individual `mlir_tablegen` commands, and finally `add_public_tablegen_target` to name that. Here, all commands will be run against `PolynomialOps.td`, and all included in the `MLIRPolynomialAttributesIncGen`.

This _accidentally_ works because `PolynomialOps.td` transitively includes other `.td` files, but we shouldn't rely on this.

j2kun replied:
Fixed in 4f8be047edc

ftynse commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,129 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_INCLUDE_MLIR_DIALECT_POLYNOMIAL_IR_POLYNOMIAL_H_
+#define MLIR_INCLUDE_MLIR_DIALECT_POLYNOMIAL_IR_POLYNOMIAL_H_
```
Nit: let's not have the `MLIR_INCLUDE` part, the rest of the codebase starts with `MLIR_DIALECT`.

j2kun replied:
Fixed in 5154bee21d4

ftynse commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,129 @@
+      : coefficient(apintBitWidth, coeff), exponent(apintBitWidth, expo) {}
+
+  Monomial(APInt coeff, APInt expo)
+      : coefficient(std::move(coeff)), exponent(std::move(expo)) {}
```
Nit: does `std::move` do something here?

j2kun replied:
clang-tidy suggested https://clang.llvm.org/extra/clang-tidy/checks/performance/unnecessary-value-param.html

But I can also resolve that by making the arguments const references, which I did in 
254d54f7ea1

ftynse commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,129 @@
+public:
+  Polynomial() = delete;
+
+  Polynomial(const llvm::SmallVector<Monomial> &terms) : terms(terms){};
```
Nit: prefer `ArrayRef<>` when taking an immutable range of things. If the callee is appending to the vector, use `SmallVectorImpl` to avoid hardcoding the number of stack elements (`SmallVector<T>` is actually `SmallVector<T, 48 / sizeof(T)>`).

Another nit: most frequently used LLVM types such as `SmallVector` and various `*Ref` are re-exported into the `mlir` namespace so there is no need to prefix them with `llvm`.

Finally: do we want to mark this constructor explicit?

j2kun replied:
Fixed in dc639f6e1ab

ftynse commented on mlir/include/mlir/Dialect/Polynomial/IR/Polynomial.h:
```diff
@@ -0,0 +1,129 @@
+};
+
+// Make Polynomial hashable.
+inline ::llvm::hash_code hash_value(Polynomial arg) {
```
Nit: `const &` on the argument?

j2kun replied:
Fixed in 77dc029158c

ftynse commented on mlir/lib/Dialect/Polynomial/IR/Polynomial.cpp:
```diff
@@ -0,0 +1,81 @@
+Polynomial Polynomial::fromCoefficients(ArrayRef<int64_t> coeffs) {
+  llvm::SmallVector<Monomial> monomials;
+  monomials.reserve(coeffs.size());
+  for (size_t i = 0; i < coeffs.size(); i++) {
```
Nit: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

j2kun replied:
Fixed in 4c030349455

ftynse commented on mlir/include/mlir/Dialect/Polynomial/IR/PolynomialAttributes.td:
```diff
@@ -0,0 +1,75 @@
+  let description = [{
+     #poly = #polynomial.poly<x**1024 + 1>
+  }];
+  let parameters = (ins "Polynomial":$polynomial);
```
Just to confirm that this works as I'd expect: does making `Polynomial` hashable ensure that `PolynomialAttr` containing equal polynomials are actually the same attribute, as in, pointing to the same impl pointer?

j2kun replied:
The generated code looks like this, which I think implies "yes"

```
struct PolynomialAttrStorage : public ::mlir::AttributeStorage {
  using KeyTy = std::tuple<Polynomial>;
...

  static ::llvm::hash_code hashKey(const KeyTy &tblgenKey) {
    return ::llvm::hash_combine(std::get<0>(tblgenKey));
  }
...
};
```

ftynse commented on mlir/lib/Dialect/Polynomial/IR/Polynomial.cpp:
```diff
@@ -0,0 +1,81 @@
+void Polynomial::print(raw_ostream &os, ::llvm::StringRef separator,
+                       ::llvm::StringRef exponentiation) const {
+  bool first = true;
+  for (const auto &term : terms) {
```
Nit: please expand `auto` unless the type is obvious from line context.

j2kun replied:
Fixed in 2cd987b843d

ftynse commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,222 @@
+    return success();
+  }
+
+  // Parse exponentiation symbol as **
+  // We can't use caret because it's reserved for basic block identifiers
+  // If no star is present, it's treated as a polynomial with exponent 1
```
Nit: please finish sentences with a period.

j2kun replied:
Fixed in 23dc5223042

ftynse commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,222 @@
+  // If no star is present, it's treated as a polynomial with exponent 1
+  if (succeeded(parser.parseOptionalStar())) {
+    // If there's one * there must be two
+    if (failed(parser.parseStar())) {
+      parser.emitError(parser.getCurrentLocation(),
+                       "exponents must be specified as a double-asterisk `**`");
```
Nit: I think `parseStar` already emits an error message, no need to emit another message.

j2kun replied:
Fixed in a1eb2b6e1ee

ftynse reviewed (changes requested):
Thanks! Requesting changes for cmake.

ftynse commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,222 @@
+/// a '+'.
+ParseResult parseMonomial(AsmParser &parser, Monomial &monomial,
+                          llvm::StringRef &variable, bool *isConstantTerm,
+                          bool *shouldParseMore) {
```
Nit: I'd rather pass these `bool`s by-reference than by-pointer since they aren't nullable.

j2kun replied:
Fixed in 16b8d968b74

ftynse commented on mlir/lib/Dialect/Polynomial/IR/Polynomial.cpp:
```diff
@@ -0,0 +1,81 @@
+namespace mlir {
+namespace polynomial {
+
+Polynomial Polynomial::fromMonomials(ArrayRef<Monomial> monomials) {
+  // A polynomial's terms are canonically stored in order of increasing degree.
+  auto monomialsCopy = llvm::SmallVector<Monomial>(monomials);
+  std::sort(monomialsCopy.begin(), monomialsCopy.end());
+  return Polynomial(monomialsCopy);
+}
```
I suspect we want to check that monomials have unique exponents and generate an error somehow when they are not.

j2kun replied:
Fixed in f6b0a6ffcaa

ftynse commented on mlir/lib/Dialect/Polynomial/IR/PolynomialAttributes.cpp:
```diff
@@ -0,0 +1,222 @@
+    }
+    monomials.push_back(parsedMonomial);
+
+    if (llvm::is_contained(exponents, parsedMonomial.exponent)) {
+      llvm::SmallString<16> coeffString;
+      parsedMonomial.exponent.toStringSigned(coeffString);
+      parser.emitError(parser.getCurrentLocation())
+          << "at most one monomial may have exponent " << coeffString
+          << ", but found multiple";
+      return {};
+    }
```
I think this should be checked in `fromMonomials` rather than in the parser.

j2kun replied:
Fixed in f6b0a6ffcaa

ftynse commented on mlir/include/mlir/Dialect/Polynomial/IR/PolynomialAttributes.td:
```diff
@@ -0,0 +1,75 @@
+  }];
+
+  let parameters = (ins
+    Builtin_TypeAttr: $coefficientType,
```
Can we use just `"Type"` here? There doesn't seem to be a need for additional boxing of a type into an attribute.

j2kun replied:
Fixed in 429f2535508

j2kun committed:
address comments about cmake config

j2kun committed:
fix include guard

j2kun committed:
remove std::move

j2kun committed:
use explicit and ArrayRef in polynomial constructor

j2kun committed:
const hash_value arg

j2kun committed:
don't evaluate size in ever loop iteration

j2kun committed:
expand auto

j2kun committed:
512 -> 16 chars smallstring

j2kun committed:
add period to end of comments, remove SmallSet include

j2kun committed:
remove parseStar error message

j2kun committed:
pass mutated bools by reference

j2kun committed:
move duplicate exponent check to fromMonomials

j2kun committed:
use Type instead of TypeAttr

j2kun committed:
clang-format

j2kun commented:
Looks like the build errors are flang on windows, unrelated to this PR. Rebasing...

joker-eph commented on mlir/include/mlir/Dialect/Polynomial/IR/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+mlir_tablegen(PolynomialTypes.h.inc -gen-typedef-decls -typedefs-dialect=polynomial)
+add_public_tablegen_target(MLIRPolynomialTypesIncGen)
+add_dependencies(mlir-headers MLIRPolynomialTypesIncGen)
+add_mlir_doc(PolynomialTypes PolynomialTypes Dialects/ -gen-typedef-doc)
```
Any reason the add_mlir_dialect() macro does not work here?

j2kun replied:
The macro assumes specific organization that I didn't use here (i.e., all tablegen in a single `Polynomial.td` file), and since I have to add calls to the underlying `mlir_tablegen` for attrdefs anyway, I felt it was simpler to keep things uniform. I also plan to upstream canonicalization patterns defined in tablegen. If you feel strongly, I could migrate this to be compatible with that macro.

joker-eph replied:
The goal of the macro I believe was also to unify the structure of all the upstream dialects somehow

j2kun replied:
Fixed in 1e6657884aa

j2kun committed:
use the add_mlir_dialect macro
